cardano-node/src/Cardano/Node/Protocol/Cardano.hs

```diff
@@ -314,11 +318,54 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
       , checkpoints = emptyCheckpointsMap
       }
 
+    -- Check that the versions for the experimental hard forks are consistent
+    warnAboutProtocolVersionsForExperimentalHardForks npcExperimentalHardForksEnabled
+      cardanoProtocolParams shelleyGenesis
+
+    return $! SomeConsensusProtocol CardanoBlockType $
+          ProtocolInfoArgsCardano cardanoProtocolParams
+
         ----------------------------------------------------------------------
         -- WARNING When adding new entries above, be aware that if there is an
         -- intra-era fork, then the numbering is not consecutive.
         ----------------------------------------------------------------------
 
+-- | Warn about the protocol versions for experimental hard forks
+warnAboutProtocolVersionsForExperimentalHardForks :: ()
+  => Bool
+  -> ProtocolParams (CardanoBlock StandardCrypto)
+  -> ShelleyGenesis StandardCrypto
+  -> ExceptT CardanoProtocolInstantiationError IO ()
+warnAboutProtocolVersionsForExperimentalHardForks npcExperimentalHardForksEnabled
+  CardanoProtocolParams
+    { hardForkTriggers = CardanoHardForkTriggers'
+        { triggerHardForkShelley = thfShelley, triggerHardForkAllegra = thfAllegra
+        , triggerHardForkMary = thfMary, triggerHardForkAlonzo = thfAlonzo
+        , triggerHardForkBabbage = thfBabbage, triggerHardForkConway = thfConway
+        }
+    } shelleyGenesis =
+  when npcExperimentalHardForksEnabled (do
+    let shelleyProtocolVersion = sgProtocolParams shelleyGenesis ^. ppProtocolVersionL
+    checkHardForkVersion "Shelley" thfShelley shelleyProtocolVersion
+    checkHardForkVersion "Allegra" thfAllegra shelleyProtocolVersion
+    checkHardForkVersion "Mary" thfMary shelleyProtocolVersion
+    checkHardForkVersion "Alonzo" thfAlonzo shelleyProtocolVersion
+    checkHardForkVersion "Babbage" thfBabbage shelleyProtocolVersion
+    checkHardForkVersion "Conway" thfConway shelleyProtocolVersion)
+  where
+    checkHardForkVersion :: ()
+      => String
+      -> Consensus.TriggerHardFork
+      -> ProtVer
+      -> ExceptT CardanoProtocolInstantiationError IO ()
+    checkHardForkVersion era (Consensus.TriggerHardForkAtVersion hfMajor) (ProtVer actualMajorVer _) =
+      let actualMajor = getVersion actualMajorVer in
+      when (hfMajor < actualMajor)
+        (liftIO $ putStrLn $ "Warning: experimental hard fork version " ++ show hfMajor ++
+         " for " ++ era ++ " is less than the actual protocol version " ++ show actualMajor ++
+         " specified for Shelley on its genesis file.")
+    checkHardForkVersion _ _ _ = return () -- No need to check for epoch-based hard forks
```

Comment:
Can you make the `_` on `Consensus.TriggerHardFork` explicit? So that, if `TriggerHardFork` receives a new case in the future, we get a non-exhaustive pattern warning?